File: Assets/Game/Scripts/Moduls/P_Modul_19.lua
local ObjectStateBackupUtils = Ballance2.Sys.Utils.ObjectStateBackupUtils

---P_Modul_19
---双向推板机关
---@class P_Modul_19 : ModulBase
---@field P_Modul_19_Flaps PhysicsObject
P_Modul_19 = ModulBase:extend()

function P_Modul_19:new()
  ModulBase.super.new(self)
  self.EnableBallRangeChecker = true
  self.BallCheckeRange = 60
end

function P_Modul_19:Active()
  ModulBase.Active(self)
  self.P_Modul_19_Flaps:Physicalize()
  self.P_Modul_19_Flaps.CollisionID = GamePlay.BallSoundManager:GetSoundCollIDByName('Wood')
end
function P_Modul_19:Deactive()
  self.P_Modul_19_Flaps:UnPhysicalize(true)
  ModulBase.Deactive(self)
end
function P_Modul_19:Reset()
  ObjectStateBackupUtils.RestoreObject(self.gameObject)
end
function P_Modul_19:Backup()
  ObjectStateBackupUtils.BackUpObject(self.gameObject)
end
function P_Modul_19:BallEnterRange()
  if self.IsActive then
    self.P_Modul_19_Flaps:WakeUp()
  end
end

function CreateClass:P_Modul_19()
  return P_Modul_19()
end

File: Assets/Game/GamePlay/Scripts/Moduls/P_Modul_19.cs
using Ballance2.Utils;
using BallancePhysics.Wapper;

namespace Ballance2.Game.GamePlay.Moduls
{
  /// <summary>
  /// 双向推板机关
  /// </summary>
  public class P_Modul_19 : ModulBase 
  {
    public PhysicsObject P_Modul_19_Flaps;

    private bool WakeUpLock = false;

    public P_Modul_19() {
      EnableBallRangeChecker = true;
      BallCheckeRange = 60;
      WakeUpLock = false;
    }

    public override void Active()
    {
      base.Active();
      WakeUpLock = false;
      P_Modul_19_Flaps.gameObject.SetActive(true);
      P_Modul_19_Flaps.CollisionID = GamePlayManager.Instance.BallSoundManager.GetSoundCollIDByName("Wood");
      P_Modul_19_Flaps.Physicalize();
    }

    public override void Deactive()
    {
      P_Modul_19_Flaps.UnPhysicalize(true);
      base.Deactive();
    }

    public override void ActiveForPreview()
    {
      gameObject.SetActive(true);
    }

    public override void DeactiveForPreview()
    {
      gameObject.SetActive(false);
    }

    public override void Reset(ModulBaseResetType type)
    { 
      WakeUpLock = false;
      ObjectStateBackupUtils.RestoreObjectAndChilds(gameObject);
    }

    public override void Backup()
    {
      ObjectStateBackupUtils.BackUpObjectAndChilds(gameObject);
    }

    protected override void BallEnterRange()
    {
      if (!IsPreviewMode && !WakeUpLock && IsActive) {
        WakeUpLock = true;
        P_Modul_19_Flaps.WakeUp();
      }
    }
  }
}